Ring buffer API changes

Mattemagikern opened the pull request:
Aims to address: #96860

Mattemagikern committed:
lib/utils: overhaul ring buffer API

This commit replaces the "ring_buf" API with a new, leaner implementation
that reduces memory footprint and improves throughput.

Note: the new API is *not* backwards compatible.
Downstream users must update to the new interfaces.

Signed-off-by: Måns Ansgariusson <Mansgariusson@gmail.com>

Mattemagikern committed:
tests/lib: add new ring_buffer API test suite

- Introduce tests exercising the new ring_buffer API
	(init, io, DMA I/O, wrap/overflow, stress, performance)
- Move existing tests for the deprecated "ring_buf" API into a
	`depricated/ringbuffer` folder so both can coexist
- Adjust CMake/Kconfig/testcase YAML to reflect the new layout

Signed-off-by: Måns Ansgariusson <Mansgariusson@gmail.com>

tmon-nordic commented on subsys/usb/device/class/cdc_acm.c:
```diff
@@ -112,8 +112,8 @@ struct cdc_acm_dev_data_t {
 	uint8_t rx_buf[CDC_ACM_BUFFER_SIZE];	/* Internal RX buffer */
-	struct ring_buf *rx_ringbuf;
-	struct ring_buf *tx_ringbuf;
+	struct ring_buffer *rx_ringbuf;
+	struct ring_buffer *tx_ringbuf;
```
For some reason this breaks my CDC ACM echo stress tester. I generally run a pseudorandom (with a known seed) sequence and expect the CDC ACM sample to echo it back. Prior to this PR, everything works fine. With this PR, after resetting the board, I keep getting mismatch always at byte 64512. I have 1024 byte long synchronization pattern before the pseudorandom sequence. 64512+1024=65536 seems to suggest that there is some sort of wrapping issue.

Mattemagikern replied:
Hi! 

Thanks for reviewing and testing!

Could you share the test with me?
I wrote this test but it didn't catch it. I tested with various size of both input & ring_buffer size.
The test should surpass the `uint16_t` overflow and would then hopefully trigger the behavior you found. 

```c 
/*
USB_CDC_ACM_RINGBUF_SIZE = 1024
#define CDC_ACM_BUFFER_SIZE (CONFIG_CDC_ACM_BULK_EP_MPS)
CDC_ACM_BUFFER_SIZE == CDC_ACM_BULK_EP_MPS = 64 or 512
*/
static uint8_t big_buff[1024];
static uint8_t big_input[66*1024];
static uint8_t big_output[66*1024];
void fill_big_input(void)
{
	sys_rand_get(big_input, sizeof(big_input));
}

ZTEST(ringbuffer_api, test_stress_big)
{
	struct ring_buffer rb;
	ring_buffer_init(&rb, big_buff, sizeof(big_buff));
	fill_big_input();

	size_t read = 0;
	size_t written = 0;
	size_t to_write;
	size_t to_read;
	uint8_t *ref;
	while (read < sizeof(big_input)) {
		to_write = MIN(ring_buffer_write_ptr(&rb, &ref), sizeof(big_input) - written);
		zassert_true(to_write > 0, "there should be space in the buffer");
		memcpy(ref, &big_input[written], to_write);
		ring_buffer_commit(&rb, to_write);
		written += to_write;

		while (read < written) {
			to_read = ring_buffer_read_ptr(&rb, &ref);
			zassert_true(to_read > 0, "there should be space in the buffer");
			memcpy(&big_output[read], ref, to_read);
			ring_buffer_consume(&rb, to_read);
			read += to_read;
		}

	}
	LOG_HEXDUMP_DBG(&big_input[64*1024], 64, "big_input");
	LOG_HEXDUMP_DBG(&big_output[64*1024], 64, "big_output");
	zassert_mem_equal(big_input, big_output, sizeof(big_input), "read != written");
}
```

tmon-nordic replied:
There are multiple ring buffers in the sample. The issue is observed on `tx_fifo.rb`. The access pattern to this ring buffer consists of 64 byte writes and 512 byte reads. I didn't yet fully analyze the failure, but the data written once is read twice.

tmon-nordic replied:
An interesting observation is that `ring_buffer_space()` underflows in the process. In the logs I get remaining space 256, 192, 128, 64, 4294966336.

tmon-nordic commented on include/zephyr/sys/ring_buffer.h:
```diff
@@ -1,13 +1,66 @@
+/* TODO: remove everything encapsulated by CONFIG_RING_BUFFER when deprecated ring buffer is
+ *	removed
+ */
+#include <zephyr/sys/ring_buffer_depricated.h>
```
Shouldn't it be `ring_buffer_deprecated.h`?

npitre reviewed (changes requested):
The commit titled "lib/utils: overhaul ring buffer API" packs too many
things together. The pipe migration should definitely be in a separate
commit.

Speaking of which, you have:

```
+               if (ring_buffer_capacity(&pipe->buf) != 0) {
+                       written += ring_buffer_write(&pipe->buf, &data[written], len - written);
```

and

```
+               if (ring_buffer_capacity(&pipe->buf) != 0) {
+                       buf.used += ring_buffer_read(&pipe->buf, &data[buf.used], len - buf.used);
```

Surely this can't be right?

I also have reservations expressed in #96860.

jfischer-no commented on subsys/usb/device/class/cdc_acm.c:
```diff
@@ -230,7 +230,7 @@ static void cdc_acm_write_cb(uint8_t ep, int size, void *priv)
 	 * timeout passes.
 	 */
-	if (ring_buf_is_empty(dev_data->tx_ringbuf) && size) {
+	if (ring_buffer_empty(dev_data->tx_ringbuf) && size) {
```
These are pointless renamings that break the API, change the code, break all the patches and OOT code for nothing. They are not acceptable. And your changes do not follow https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#pr-requirements, please get more familiar with development process first.

Mattemagikern committed:
kernel: pipe: Update to use ring buffer API

Just for tests to run.

Signed-off-by: Måns Ansgariusson <mansgariusson@gmail.com>

Mattemagikern committed:
net: lib: sockets: select RING_BUFFER for socketpair

The socketpair implementation uses the ring buffer for data storage.
The kconfig should select RING_BUFFER for socketpair to allow building
without implied dependencies.

Signed-off-by: Måns Ansgariusson <mansgariusson@gmail.com>

tmon-nordic commented:
> possible sed script

Why use bash + sed when there is [Coccinelle](https://coccinelle.gitlabpages.inria.fr/website/) that is better suited for the job?

Mattemagikern committed:
coccinelle: ring_buf -> ring_buffer & api breaking changes

loop_check.cocci: finds loops where the ring_buf_[put|get]_claim and
ring_buf_[put|get]_finish functions are used in different scopes (claim in a
loop, finish outside the loop).

ring_buf.cocci: updates the ring_buf API to ring_buffer API. Changes the
function names & modifies the function parameters to match the new API.
Note: this does not fix macros.

Signed-off-by: Måns Ansgariusson <mansgariusson@gmail.com>
